### ISSUE TITLE
✍️ Contribution period: <Ashita Srivastava> 

### ISSUE THREAD

**Issue by Ashita17** [ersilia-os/ersilia#1041]
### Week 1 - Get to know the community

- [X] Join the communication channels
- [X] Open a GitHub issue (this one!)
- [x] Install the Ersilia Model Hub and test the simplest model
- [x] Install Docker if needed, and test another model
- [x] Write a motivation statement to work at Ersilia
- [x] Submit your first contribution to the Outreachy site

### Week 2 - Get Familiar with Machine Learning for Chemistry

- [x] Select a model from the list suggested in GitBook
- [x] Download and serve the model via the Ersilia Model Hub to ensure it works
- [x] Open a repository on your GitHub user with all the necessary files
- [x] Select and clean a dataset of 1000 molecules (example notebook 1)
- [x] Run predictions for the molecules on the selected model and evaluate the results

### Week 3 - Validate a Model in the Wild

- [x] Find a suitable dataset with sufficient experimental results
- [x] Clean and standardize the dataset
- [x] Run predictions and calculate metrics.

### Week 4 - Prepare your final application

- [ ] Submit the final application in the Outreachy website

**Comment by GemmaTuron** [comment]
Hi @Ashita17 

We are on week 3 of the contribution period. Please let us know if you plan to continue with your contribution by the next 2 days, otherwise we will close this issue so we can focus on the applicants who want to make a final application to Ersilia.

**Comment by Ashita17** [comment]
I want to continue, I was just having some problems in the task 2 of week , will post the GitHub repo link soon for the deepherg model that I am currently working on

**Comment by Ashita17** [comment]
I wrote a motivation statement in the outreach contribution itself. However, I can see that people have written it on their issues. Sorry for the misunderstanding. Please review it here as well. @GemmaTuron @DhanshreeA 

**MOTIVATION STATEMENT**

I wanted to join outreachy to work with a community of supportive mentors and peers which is missing in my college where technical clubs are more or less a gendered space ,which believe that only boys have good analytical and mathematical ability and girls are not at par with them. I am the one of the two girls in the data science club of my college out of 60 people there. I have a strong mathematical background as I rigorously prepared for and cracked the JEE exam in India.

I chose Ersilia because of the nature of project. I was particularly interested in doing an ML project where I can learn to run ML models efficiently. I have done some quality data science projects in my college club and its always fun to tackle new issues that come along especially when we are trying to improve the dataset such as making the frequency of the inputs more uniform, etc or when we are trying to improve the ML model by changing the various parameters and sometimes ending up employing a new method altogether in the end !

Through this contribution phase and hopefully the internship, I want to improve my skills to become a proficient data scientist who is good at handling data as well as developing new models. I particularly like machine learning because it is much more analytical and teaches one patience owing to its rigorous mathematical background. I read “*Hands-On Machine Learning with Scikit-Learn, Keras, and TensorFlow***”** and it deepened my curiosity to explore this field further.

After this internship, I plan to work as a quantitative research analyst in a renowned firm like Goldman Sachs or JPMorgan Chase & Co. to gain experience of real life projects in an environment of highly skilled individuals and maybe after a few years go for higher studies in machine learning. Most importantly, I would want to encourage young girls to put themselves out there in male dominated fields, most of which require mathematical and analytical ability. I would like to volunteer to teach them maths or the concepts of ML once I have the necessary skills.

**Comment by Ashita17** [comment]
Sorry for adding the adding the work late, I misunderstood that everything has to be attached in the form of contribution on the outreachy website.

**Comment by Ashita17** [comment]
In week 1, I learned how to work with Conda environments and managing packages. I faced various errors while trying to run the ersilia model such as difficulty in relative imports, container is not ready yet, etc. However after trying to fetch the model multiple times , I successfully downloaded ersilia and fetched the eos3b5e model as well.
<img width="700" alt="Screenshot 2024-03-18 at 10 57 17 PM" src="https://github.com/ersilia-os/ersilia/assets/101327935/6ccfa5cd-be34-48bb-8563-bb751e133f5f">

**Comment by Ashita17** [comment]
## Week 2

I have chosen the model eos30f3. I successfully fetched and served the model and ran predictions on the 1000 molecules given in the reference_library.csv file. I halfway done with week 2 task 1. I have to evaluate the model and check its reproducibility to complete week 2 tasks.

**Comment by GemmaTuron** [comment]
Hi @Ashita17 

Can you provide a short summary of week 2 tasks when you complete them so we can provide feedback?

**Comment by Ashita17** [comment]
Definitely, I have been facing some issues in the task 2 of week two but I' ll share the summary very soon . Sorry for falling behind the schedule and thank you for your concern, it means a lot.

**Comment by Ashita17** [comment]
## Week 2 Update

There were two sets of models we were supposed to choose from : hERG models and ADME models. I was very excited to experiment with the hERG models. Why? because they talked about cardiotoxicity! the biology subject area which excited me the most when I was in school !!

I initially chose the model eos30gr (deepherg) simply because its publication paper has beautifully explained about the multi-task deep neural network and it looked really fun to look into the Mol2vec Descriptors and MOE Descriptors and get to know more and more about both chemistry and machine learning. However, after trying numerous times, it kept giving me the same null input, I wasn't able to resolve the error even after my peers helped me on slack for which I am really grateful😇.
<img width="1214" alt="Screenshot 2024-03-11 at 12 19 48 AM" src="https://github.com/ersilia-os/ersilia/assets/101327935/83780e19-349f-4188-8d3b-7974279e4a0c">

I finally decided to randomly try some other model and tested eos30f3 (dmpnn-herg). This model was working fine for all the smiles and giving outputs but there was something weird I noticed when I finally started to evaluate the results I ran on the reference library shared with us on slack.
<img width="572" alt="Screenshot 2024-03-21 at 9 14 42 PM" src="https://github.com/ersilia-os/ersilia/assets/101327935/a689bf13-aa82-4e19-b469-6295cb57d825">
It showed that the number of blockers in the dataset is much more than the number of non-blockers which is very unlikely. I couldn't trust its results and started to go through all the models again.

I was on the verge of giving up but I finally stumbled upon BayeshERG model (eos4tcc). The interesting thing about this model is that while the other models just give the hERG blocking activity , it gives two more parameters : epis (epistemic probability)and alea (aleatoric probability). As we can guess from the name, this model deals with not only predictions but also the uncertainities that may come along. I have a deep interest in statistics and prbability and  surely, this was THE model for me!

Its a bayesian graph neural network model.

<img width="1252" alt="Screenshot 2024-03-22 at 5 45 29 AM" src="https://github.com/ersilia-os/ersilia/assets/101327935/dec85997-8592-490f-9fb2-c934a2654872">

## Understanding of the model 💭

The model takes 1000 smiles as input from the dataset "1000_values_dataset.csv". It generates inchikey and gives the following three outputs:

1. score: This is the hERG channel inhibition percentage at 10µm. BayeshERG is a graph neural network (GNN)-based Bayesian DL model. It calculates the hERG blocking probability using the softmax function. If the score is ≥ 0.5, the molecule is calssified as a hERG blocker and if the score is < 0.5 it is considered to be a non-hERG blocker. This score helps us to perform classification labelling the blackers as 1 and non-blockers as 0
2. alea: This refers to the "aleatoric" uncertainity which is referred to as the model uncertainity due to the lack of data.
3. epis: This refers to the "epistemic" uncertainity which is referred to as the model uncertainity due to data’s intrinsic randomness which may be caused due to noise inherent in the dataset.

Comment on uncertainity : The uncertainity or VARIANCE is the sum total of quantified epistemic and aleatoric uncertainty.

<img width="474" alt="variance_formual" src="https://github.com/ersilia-os/ersilia/assets/101327935/85de6d6b-ff3a-4fc8-9988-3d55edfaffd7">


I successfully fetched the model, served it and it gave me predictions in one go 🥹
<img width="712" alt="Screenshot 2024-03-22 at 5 53 15 AM" src="https://github.com/ersilia-os/ersilia/assets/101327935/1cf1b179-2253-42ae-bee7-65bbb32e2b32">


I went on to develop these results from the model:-

<img width="579" alt="Screenshot 2024-03-22 at 5 55 26 AM" src="https://github.com/ersilia-os/ersilia/assets/101327935/b7d20d05-c79a-404c-a7f2-b53bbc27e7ef">

#### Inference from the histogram
1. Most of the molecules (about 25%) have hERG blocking activity score in the range (0.1-0.2)
2. Molecules with very strong hERG blocking activity i.e. hERG blocking activity score > 0.9 are just approximately 1%.
3. Unexpected result: I did not expect that there will be molecules in the score range (0.6-0.7) and (0.8-0.9) but suprisingly that's the case!

<img width="642" alt="Screenshot 2024-03-22 at 5 56 55 AM" src="https://github.com/ersilia-os/ersilia/assets/101327935/0346fec1-6e89-482f-b035-0da7e3dc8c01">

#### Inferences from the pie chart 
1. We can see that more than 50 percent of the molecules have hERG blocking activity score less than 0.4 which means that they are non-blockers.
2. This data is very well in agreement with any random dataset that we come across in real life where the number of non-hERG blockers dominates the number of hERG blockers
3. Let's go ahead and make this visualisation even better!

<img width="678" alt="Screenshot 2024-03-22 at 5 57 58 AM" src="https://github.com/ersilia-os/ersilia/assets/101327935/d06c5a28-1805-42dd-a521-ca19cd43fc97">

### Let's dive into uncertainities ! 💡

As discussed before, aleatoric uncertainity arises from inherent randomness of the data and this component of uncertainity can't be reduced even by gathering more information. On the other hand , epistemic unceratinity arises due to lack of information and therefore can be reduced by gathering more information. In this section, I would like to analyse the data for aleatoric and epistemic unceratinities to know if the model is lagging due to lack of information.

<img width="625" alt="Screenshot 2024-03-22 at 5 58 39 AM" src="https://github.com/ersilia-os/ersilia/assets/101327935/e22eaa20-edd9-48d6-af5e-f2cd88e8a679">

### Inference from the line graphs 

1. From the above graph, we can see that the maximum uncertainity arises when we are in the zone of hERG blocking activity score values between 0.4 to 0.6 which means the predictions are least trused when they are made for molecules lying in this range.
2. Scope of improvement: we can see from the above graph that although the variance/uncertainity is quite high for the range 0.4-0.6, most of it comes from the aleatoric component which is the non-reducible part of uncertainity as discussed before. Only a very small portion of uncertainity comes from the epistemic part (reducible part) but this portion is quite insignificant when looking at the total variance/uncertainity.
3. Thus, the model performs very well and most of the uncertainity it faces is due to the inherent randomness present in data which cannot be ruled out.
4. If improvements have to be made, they should be made for molecules in the score range 0.4-0.8 where the epistemic component forms maximum percentage of total variance. This can be done by collecting more data for this range. Since our main culprit is the epistemic component ,let's further analyse this component using scatter plots!

### Scatter Plot - Veryfying our observations for aleatoric uncertainity !

<img width="609" alt="Screenshot 2024-03-22 at 5 59 51 AM" src="https://github.com/ersilia-os/ersilia/assets/101327935/87c8cb6b-f4e6-4bdb-9a7f-09c898d6ae29">

#### Inferences from scatter plot

1. Clearly we can see that we have a lot of data points in the range 0.0-0.4 and here the aleatoric uncertainity is very less. 
2. However, as we move to the 0.4-0.8 score range, the number of data points decreases and the aleatoric uncertainity increases due to lack of data.
3. For the 0.8-1.0 range, very few molecules are present so solid conclusions can't be made!

**Comment by Ashita17** [comment]
@GemmaTuron @DhanshreeA , I have completed task 1 of week 2 and currently working on task 2 of week 2. Kindly review my tasks and suggest changes if any . Posting the link to the repo again [here](https://github.com/Ashita17/Ersilia-Outreachy/tree/main). Thank you :) .

**Comment by DhanshreeA** [comment]
@Ashita17 good work so far, you are in the right direction with Task 1. However I would recommend creating a final application starting next week and linking this issue there. Please do not attempt task 3 as we would not be able to provide feedback on that at this point. I can review your Task 2 work on Monday.

**Comment by Ashita17** [comment]
@DhanshreeA I had already been reading about task 3 and did some work. I understand if you won' t be able to provide feedback due to the time crunch. Since, I had been working on task 3, can I just mention the work I the part of task 3 that I am able to do till today EOD on this issue?

**Comment by Ashita17** [comment]
# Week 2 Update 

## Task 2 - Reproducibility

For this task we were supposed to do the following:
- Select a dataset to reproduce the result of the model through original implementation: Here, I have taken [EX4.csv](https://github.com/GIST-CSBL/BayeshERG/blob/main/data/External/EX4.csv) which was the external dataset used by the authors of the publication for testing the BayeshERG model.
- Identify a result you could reproduce from the publication: Since the model deals not only with the predictions but also the uncertainities that come along, I have decided to compare the hERG blocking activity score , epistemic and aleatoric unceratainity as well as total variance of the publication implementation (BayeshERG) vs the model fetched from ersilia (eos4tcc).
- Check that the model provides the same results when running via the Ersilia Model Hub: In the last section , I will compare the results obatined through the two methods.

### *Setting up BayeshERG on my system*
This was the longest process. I use MacBook Air M1 and the model specifically required CUDA , which is not supported by the MacM1 silicon chips, so I decided to work on colab. There as well, it was difficult to create a Conda environment and the terminal can be used only by users that have pro-access. So finally , I decided to use a friend's laptop. Here also I made some syntax errors again and again while following the instructions but finally I saw this on my screen and I was on cloud nine!
![WhatsApp Image 2024-03-23 at 14 14 49](https://github.com/ersilia-os/ersilia/assets/101327935/57df3b82-b41b-46c7-8b7f-85ffa1d807c0)
I had set the sampling time to 100 .The default time that the model mentions on the site is 30 but since the authors used 100 as the sampling time while testing the model, so I chose 100 as well. 

I proceeded on to make comparisons between the BayeshERG model and eos4tcc model and came up with the following analysis.

## Comparison between predictions of BayeshERG and eos4tcc Models

### Bar Chart 📊

![bar_graph](https://github.com/ersilia-os/ersilia/assets/101327935/32cc3c8f-5d26-4e15-b7c5-f9ce91436c89)

### Inferences

Similarity: 
1. Relative number of hERG and non-hERG blockers: We can see that in both the predictions, the number of non-hERG blockers (hERG blocking activity score < 0.5) is greater than the number of hERG-blockers (hERG blocking activity score >= 0.5).

Differences:
1. BIAS: BayeshERG model gives an appreciable number of strong hERG blockers (hERG blocking activity score > 0.8) but the eos4tcc model reports no strong hERG blockers. We can say that eos4tcc is biased towards making non-hERG predictions and hardly reports the strong hERG blockers. Also , it has a bias towards reporting very weak-hERG blocking activity as it significantly reports much more molecules in the lower range of hERG blocking activity score (0-0.2) in comparison to BayeshERG.

Let's look at pie charts for even better visualisation !

### Pie Charts 🟡

![pie_charts](https://github.com/ersilia-os/ersilia/assets/101327935/6d230315-486d-41ef-9f52-068adc96ec62)

### Inferences

Similarity: 
1. Both the models report approximately 65% non-blockers and 35% blockers

Differences:
1. BIAS: We can see that there are hardly any strong blockers for eos4tcc model (red  and orangered zones) but they do exist in the BayeshERG model. Thus, eos4tcc is biased towards reporting molecules as non-hERG blockers and hardly reports strong hERG-blockers.
2. Percentage Difference : In terms of overall predictions, eos4tcc detects 14.7% less hERG blockers in comparsion to the original publication model BayeshERG.

### Comparing the uncertainities 🤔

![uncertainity](https://github.com/ersilia-os/ersilia/assets/101327935/0cabde9d-c3e7-4629-9945-ebab85abe2f2)

### Inferences

There are significant differences when it comes to the uncertainities.

1. Overall variance: We can see that the overall variance almost follows a normal distribution for the BayeshERG model whereas for the eos4tcc model, it increases in the beginning, reaches a peak value and then falls. However, there is a similarity in terms of the maximum value of variance which for both the models remains around 0.25.
2. Epistemic uncertainity: It remains almost constant for both the models but for the eos4tcc model, it is lower . This means that the model is under-reporting the epistemic uncertainity.
3. Aleatoric uncertainity: It varies just like the variance, follows a somewhat normal distribution for BayeshERG model and increases in the beginning, reaches a peak value and then falls for the eos4tcc model.

## Final Comment on Reproducibility ✔️

Since the model focusses on both predictions as well as uncertainities let's talk about both!

1. Predictions: The models eos4tcc and give similar results but the error in reporting hERG blocker is 14.7% which an appreciable value. eos4tcc is hardly able to predict any strong hERG blockers whereas BayeshERG does whihc is a strong feature of BayeshERG as mentioned in the publication.
2. Uncertainities: When it comes to the overall uncertainity/variance values, the models show significant difference in terms of the pattern they follow although the maximum variance reported by both the model remians about 0.25. The epistemic uncertainity doesn't vary much and the aleatoric uncertainity folows a pattern very similar to the variance.

THUS THE MODEL IS REPRODUCIBLE BUT NEEDS A LOT OF IMPROVEMENT AS APPRECIABLE DIFFERENCES EXIST!

**Comment by Ashita17** [comment]
@DhanshreeA @GemmaTuron kindly review my task 2 for week 2. putting the link [here](https://github.com/Ashita17/Ersilia-Outreachy/blob/main/model-validation-example/notebooks/01_model_reproducibility.ipynb) again for convinience, thank you !

**Comment by Ashita17** [comment]
# Week 3 Update

**Comment by Ashita17** [comment]
@DhanshreeA @GemmaTuron please find the link to the notebook for week 3 [here](https://github.com/Ashita17/Ersilia-Outreachy/blob/main/model-validation-example/notebooks/02_external_validation.ipynb) . You asked me not to attempt task 3 but since I had already started reading about and working on it, I was really excited to try it out as I still had quite some time left. It is understandable if you can't review it due to shortage of time but if there's any chance you can review it, it would mean a lot to me. Thank you :).

**Comment by DhanshreeA** [comment]
@Ashita17 you've done a very good job! Thank you for this work, it's quite helpful for us to go and review this model's implementation. It is likely a retrained model thus explaining the discrepancies in results. Please go ahead and submit your final application!

Just a small comment for the future, Pi charts are not a great idea for visualizing probability distributions. Everything else looks great! :)

**Comment by Ashita17** [comment]
Thank you so much for your appreciation ma'am, you made my Holi 10x better !! I'll keep the pie chart point in mind 👍🏼

**Comment by Ashita17** [comment]
@DhanshreeA , I have a doubt. You said pie charts are not a great way of visualising probability distributions . Shall I remove all the pie charts from week 2 and week 3 tasks then or you were talking about certain sections of the tasks? I am a bit confused , could you clarify please?
